low-level/framework/utils/iem.py

```diff
@@ -0,0 +1,140 @@
+            "Started server local drives monitoring.",
+            ""],
+        "0050010005" : ["Smartctl command execution error.",
+            "Can not fetch drive information.",
+            "Reinstall/reconfigure smartmonotool package."],
+        "0050010006" : ["Smartctl command executed successfully.",
+            "Can fetch drive information.",
+            ""],
+        "0050010007" : ["UDisks2 is not installed.",
+            "Can not fetch drive information using dbus interface.",
+            "Reinstall/reconfigure UDisks2 package."],
+        "0050010008" : ["UDisks2 is available.",
+            "Can fetch drive information using dbus interface.",
+            ""],
+        "0050020001" : ["Kafka service is not in active state.",
+            "SSPL can not raise an alerts.",
+            "Reconfigure/start kafka service."],
+        "0050020002" : ["Kafka service is in active state.",
+            "Kafka service is available.",
+            ""]
+    }
+
+    def check_previous_iem_event(self, event_name, event_code):
+        """Before logging iem, check if is already present."""
+        previous_iem_event = None
+        is_iem_exist = False
+        if not os.path.exists(IEM_DATA_PATH):
+            os.makedirs(IEM_DATA_PATH)
+        iem_event_path = f'{IEM_DATA_PATH}/iem_{event_name}'
+        if not os.path.exists(iem_event_path):
+            with open(iem_event_path, 'w') as f:
+                f.write(event_code)
+                f.close()
+        else:
+            with open(iem_event_path, 'r') as f:
+                previous_iem_event = f.read().strip()
+                if previous_iem_event != event_code:
+                    with open(iem_event_path, 'w') as file:
+                        file.write(event_code)
+                        file.close()
+                else:
+                    logger.info("%s - IEM already created." %event_code)
+                    is_iem_exist = True
+                f.close()
+
+        return is_iem_exist
+
+    def check_fault_event(self, event_name, *events):
+        """Before logging fault_resolved iem event,
+        Check if fault iem event is present for that particular event."""
+        fault_iem = False
+        iem_event_path = f'{IEM_DATA_PATH}/iem_{event_name}'
+        if os.path.exists(iem_event_path):
+            with open(iem_event_path, 'r') as f:
+                previous_iem_event = f.read().strip()
+            if previous_iem_event in events:
+                fault_iem = True
+        return fault_iem
+
+    def generate_iem(self, severity, event_code, event_name, description):
+        """Generate iem."""
+        iem = "IEC:%sS%s:%s" %(severity, event_code, description)
+        previous_iem = self.check_previous_iem_event(event_name, event_code)
+        if not previous_iem:
+            log_iem(iem)
+
+    def create_iem_fields(self, event, severity):
+        event_code = event[0]
+        event_name = event[1]
```

Comment:
what event name we get usually, give me some examples. am trying to understand the filename the persistence cache will have too eventually, with thsi event name

Reply:
here we are fetching event_name from "EVENT_CODE" dictionary, which is defined in same class.
event names will be IPMITOOL, HDPARM, SMARTCTL, UDISKS2, and KAFKA.
persistence file name - iem_IPMITOOL, iem_SMARTCTL etc.